### PR TITLE
Ruby: Decompression Bombs

### PR DESCRIPTION
**Pull request by am0o0**
This is part of All for one, one for all query submission, I'm going to submit an issue in github/securitylab for this pull request too.
I wrote two versions which V2 is using recursion by use of API graphs to find zip file sinks. I added V1 because I think it maybe have some advantage in your opinion but I prefer V2. the predicates in V2 can used in ZipSlip query too I can work on this too if you agree.( I couldn't write the recursion predicate without the initial help of @asgerf in #13431)
I added the only sanitizer that I think is available which is checking the size of each zip entry.
Also I added some instance of some really popular additional taint steps which I think it is good to be added as global steps.

### CODE REVIEW

**Comment by smowton** [comment]
Anyone not watching the repo in general, note this is part of a family of submissions:

https://github.com/github/codeql/pull/13553
https://github.com/github/codeql/pull/13554
https://github.com/github/codeql/pull/13555
https://github.com/github/codeql/pull/13556
https://github.com/github/codeql/pull/13557
https://github.com/github/codeql/pull/13558

**Comment by calumgrant** [comment]
Hey @amammad, would you be able to submit these to the [Security Lab bug bounty program](https://securitylab.github.com/bounties/) using [this template](https://github.com/github/securitylab/issues/new?template=all-for-one.yml) as this will help us prioritise this work, and it would also be really helpful to test the CVEs found as well.

Thank you for your submission!

**Review by alexrford — CHANGES_REQUESTED** [review]
Hi, I've run an initial review of this. Some of the feedback here may overlap with comments on the versions of this query for other languages.

**Review by alexrford — APPROVED** [review]
Sorry for the delay, approved.

File: ruby/ql/src/experimental/CWE-522-DecompressionBombs/DecompressionBombs.qhelp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+<p>Extracting Compressed files with any compression algorithm like gzip can cause to denial of service attacks.</p>
+<p>Attackers can compress a huge file which created by repeated similiar byte and convert it to a small compressed file.</p>
+
+</overview>
+<recommendation>
+
+<p>When you want to decompress a user-provided compressed file you must be careful about the decompression ratio or read these files within a loop byte by byte to be able to manage the decompressed size in each cycle of the loop.</p>
+
+</recommendation>
+<p>Please read official RubyZip Documentation <a href="https://github.com/rubyzip/rubyzip/#size-validation">here</a></p>
```
Comment (alexrford) [inline ruby/ql/src/experimental/CWE-522-DecompressionBombs/DecompressionBombs.qhelp:15]:
```suggestion
<p>Please read official RubyZip Documentation <a href="https://github.com/rubyzip/rubyzip/#size-validation">here</a></p>
</recommendation>

```

This should fix the qhelp check error.

File: ruby/ql/src/experimental/CWE-522-DecompressionBombs/DecompressionBombs.ql [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @name User-controlled file decompression
+ * @description User-controlled data that flows into decompression library APIs without checking the compression rate is dangerous
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.8
+ * @precision high
+ * @id rb/user-controlled-file-decompression
```
Comment (alexrford) [inline ruby/ql/src/experimental/CWE-522-DecompressionBombs/DecompressionBombs.ql:8]:
```suggestion
 * @id rb/user-controlled-data-decompression
```

For consistency with JS and to address the duplicate query ID check.

File: ruby/ql/test/query-tests/experimental/CWE-522-DecompressionBombs/DecompressionBombs.expected [hunk 3]
```diff
@@ -0,0 +1,114 @@
+edges
```
Comment (alexrford) [inline ruby/ql/test/query-tests/experimental/CWE-522-DecompressionBombs/DecompressionBombs.expected:1]: This `.expected` file looks like it needs to be regenerated to account for a change in the `edges` relation format.

**Review by alexrford — CHANGES_REQUESTED** [review]
Just a couple of suggestions to make CI pass.

**Review by alexrford — APPROVED** [review]

**Pull request merged**